rust-bitcoin-derive/src/lib.rs

```diff
@@ -0,0 +1,87 @@
+//! Custom derive macros for the rust-bitcoin project.
+//!
+
+extern crate proc_macro;
+extern crate syn;
+extern crate quote;
+
+use proc_macro::TokenStream;
+use syn::{parse_macro_input, Data, DeriveInput, Fields};
+use quote::quote;
+
+#[proc_macro_derive(Encodable)]
+pub fn encodable_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let ident = input.ident;
+    let data = input.data;
+
+    let fields = match data {
+        Data::Struct(ref data) => {
+            match data.fields {
+                Fields::Named(_) => &data.fields,
+                Fields::Unnamed(_) | Fields::Unit => {
+                    panic!("Encodable only implemented for named structs");
+                }
+            }
+        }
+        Data::Enum(_) | Data::Union(_) => panic!("Encodable only implemented for structs"),
+    };
+
+    let consensus_encode_fields = fields.iter().map(|f| {
+        let ident = f.ident.clone().unwrap();
+        quote! {
+            len += self.#ident.consensus_encode(&mut s)?;
+        }
+    });
+
+    let gen = quote! {
+        impl Encodable for #ident {
```

Comment:
At first I was using absolute paths but I couldn't get them to work, two reasons jump back to my mind:
1. `io` doesn't work because of how we conditionally bring in `io` from either `core2` or `std` depending on `no-std` feature.
2. The bitcoin types can't be brought into scope because one cannot use `bitocin::encode::Encodable` inside the bitcoin crate?

Reply:
> At first I was using absolute paths but I couldn't get them to work, two reasons jump back to my mind:
> 1. `io` doesn't work because of how we conditionally bring in `io` from either `core2` or `std` depending on `no-std` feature.

Having `Encodable` and `Decodable` be no-std friendly (with home-grown io traits) might be enough of a reason to move it all into its own crate? I think the home-grown io traits have been tried before but I don't recall what didn't work.

> 2. The bitcoin types can't be brought into scope because one cannot use `bitocin::encode::Encodable` inside the bitcoin crate?

Yeah that is what I meant with "use `crate::`" but that makes them unusable outside of `bitcoin`.

You could have a proc-macro parameter (like `serde` has one for when you rename the dependency) to configure the import crate but given that this problem is also solved by moving them into a separate crate, it doesn't seem worth it.